docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+- `Vectors`
+    - `_view_indices: Array[int, (-1)]`, indices of vectors that are in the current slice.
+        - lots of private properties derived from this (e.g. `_view_face_color`, `_view_faces`).
+
+The vispy layers also read other state from their corresponding layer.
+In particular they read `Layer._transforms` to produce a transform that can be used to properly
+represent the layer in the vispy scene coordinate system.
+
+## Detailed description
+
+The following diagram shows the new proposed approach to slicing layers asynchronously.
+
+![](https://raw.githubusercontent.com/andy-sweet/napari-diagrams/main/napari-slicing-async-calls.drawio.svg)
+
+As with the existing synchronous slicing design, the `Dims.events.current_step` event
+is the shared starting point. In the new approach, we pass `ViewerModel.layers` through
+to the newly defined `LayerSlicer`, which makes a slice request for each layer on the main thread.
+This request is processed asynchronously on a dedicated thread for slicing, while the main thread
+returns quickly to the Qt main event loop, allowing napari to keep responding to other updates
+and interactions. When all the layers have generated slice responses on the slicing thread,
+the `slice_ready` event is emitted. That triggers `QtViewer.on_slice_ready` to be executed on
+the main thread, so that the underlying `QWidgets` can be safely updated.
+
+The rest of this section defines some new types to encapsulate state that is critical
+to slicing and some new methods that redefine the core logic of slicing.
+
+### Request and response
+
+First, we introduce a request to encapsulate the required input to slicing.
+
+```python
+class LayerSliceRequest:
+    data: ArrayLike
+    world_to_data: Transform
+    point: Tuple[float, ...]
+    dims_displayed: Tuple[int, ...]
+    dims_not_displayed: Tuple[int, ...]
+```
+
+At a minimum, this request must contain the data to be sliced, the point at
+which we are slicing in napari's shared world coordinate system, and a way to
+transform world coordinates to a layer's data coordinate system.
+In the future, the point may become a bounding box that includes the current
+field of view.
+Given our understanding about the existing slice input state, each layer type
+should extend this request type to capture the input it needs for slicing, such
+as `Points.face_color`.
+
+Second, we introduce a response to encapsulate the output of slicing.
+
+```python
+class LayerSliceResponse:
+    data: ArrayLike
+    data_to_world: Transform
+```
+
+At a minimum, this response must contain the sliced data and a transform that
+tells us where that data lives in the currently displayed scene of napari's
+shared world coordinate system.
+At this point `data` should be fast to access either from RAM (e.g. as a numpy
+array) or VRAM (e.g. as a CuPy array), so that it can be consumed on the main
+thread without blocking other operations for long.
```

Comment:
I think this requires some extra thought; I'm trying to imagine how this all would work out in a possible future where we have `DataSources` that maintain data loaded onto the GPU so that we can for example share visualizations between canvases, or between layers, or allow faster slicing in the shader. Indeed, I'm working towards this in PRs such as vispy/vispy#2350 and vispy/vispy#2351. Is there a way we can take advantage of this async machinery, while also allowing "pre-loading" data onto the GPU and do the slicing in the shader?

Reply:
Slicing in the shader is super interesting. I suspect that will need to be in addition and complementary to this PR, so you could say have a 4D timeseries dataset that you sliced to get a smaller single 3D timepoint that you "pre-loaded" (as you say above) and then sliced in the GPU to get a specific 2D plane. Does it sound like we're on the same page here @brisvag?

Reply:
Yup :) But I think the biggest advantage would be for multicanvas in order to display the same data multiple times using the same texture; imagine having a sizeable 3D volume on which you performed feature picking in 10 different ways, and you'd like to see them side by side. You could have a grid of 10 canvases with linked cameras and zoom in onto details and compare, but without sharing texture you'd be decuplicating the vram usage.

Reply:
This might be an oversimplification, but as long as the thing that's rendering the slice (e.g. vispy) has enough to be able to do that "quickly", then I think the job of slicing as described here is done. In the case described in the comments above (i.e. 4D data, same underlying 3D slice with different features presented in multiple canvases), I'd imagine that slicing just wouldn't do much after the 3D sliced data has been read for the first time. Not quite sure what `data` would be in that case, which is one motivation to keep the typing of that as vague as possible for now (e.g. `Any`) and maybe keep things temporarily private.